<a href="https://colab.research.google.com/github/francomanca93/analisis-de-datos/blob/aplicaciones/3-Aplicando-pandas/16_Group_by_Extrayendo_valor_con_variables_categ%C3%B3ricas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Group by - Refozando el uso de esta función

En esta sección veremos nuevamente Group by y rezorzaremos con otro dataframe el uso de esta función. Veremos como agrupar nuetro dataframe bajo una categoria y sobre esta aplicar algunas funciones estadísticas. 

In [43]:
import pandas as pd
import numpy as np

In [44]:
pd.options.display.float_format = '{:,.2f}'.format

In [45]:
import seaborn as sns

El dataset sobre el que trabajaremos es sobre **propinas**. En él aparecen parámetros sobre la cuenta, la propina que dieron las personas, si fue un hombre o mujer, si era fumador, tambien si como era el día (soleado, nublado, etc), el horario de la misma (desayuno, almuerzo, cena, etc) y cuantas personas compartian la comida. 

In [46]:
df = sns.load_dataset('tips')
df

total_bill  tip     sex smoker   day    time  size
0         16.99 1.01  Female     No   Sun  Dinner     2
1         10.34 1.66    Male     No   Sun  Dinner     3
2         21.01 3.50    Male     No   Sun  Dinner     3
3         23.68 3.31    Male     No   Sun  Dinner     2
4         24.59 3.61  Female     No   Sun  Dinner     4
..          ...  ...     ...    ...   ...     ...   ...
239       29.03 5.92    Male     No   Sat  Dinner     3
240       27.18 2.00  Female    Yes   Sat  Dinner     2
241       22.67 2.00    Male    Yes   Sat  Dinner     2
242       17.82 1.75    Male     No   Sat  Dinner     2
243       18.78 3.00  Female     No  Thur  Dinner     2

[244 rows x 7 columns]

In [47]:
df.describe(include='all')

total_bill    tip   sex smoker  day    time   size
count       244.00 244.00   244    244  244     244 244.00
unique         nan    nan     2      2    4       2    nan
top            nan    nan  Male     No  Sat  Dinner    nan
freq           nan    nan   157    151   87     176    nan
mean         19.79   3.00   NaN    NaN  NaN     NaN   2.57
std           8.90   1.38   NaN    NaN  NaN     NaN   0.95
min           3.07   1.00   NaN    NaN  NaN     NaN   1.00
25%          13.35   2.00   NaN    NaN  NaN     NaN   2.00
50%          17.80   2.90   NaN    NaN  NaN     NaN   2.00
75%          24.13   3.56   NaN    NaN  NaN     NaN   3.00
max          50.81  10.00   NaN    NaN  NaN     NaN   6.00

La variable **day** tiene 4 categorias. Veamos a esas 4 categorias y contemos cuantas veces se repiten cada una. 

In [48]:
df['day'].value_counts()

Sat     87
Sun     76
Thur    62
Fri     19
Name: day, dtype: int64

Ahora veamos en % los numeros anteriores. Lo que hacemos es dividir la suma de los valores a cada valor y multiplicarlo por 100.

In [49]:
df['day'].value_counts() / df['day'].value_counts().sum() * 100

Sat    35.66
Sun    31.15
Thur   25.41
Fri     7.79
Name: day, dtype: float64

Podemos tener estimaciones estadisticas por alguna de nuestras categorias. 

In [50]:
df.groupby('sex').mean()

total_bill  tip  size
sex                          
Male         20.74 3.09  2.63
Female       18.06 2.83  2.46

Veamos en porcentaje cuanto corresponde la propina respecto al total de la cuenta.

In [51]:
df['porcet_tip'] = df['tip'] / df['total_bill']
df

total_bill  tip     sex smoker   day    time  size  porcet_tip
0         16.99 1.01  Female     No   Sun  Dinner     2        0.06
1         10.34 1.66    Male     No   Sun  Dinner     3        0.16
2         21.01 3.50    Male     No   Sun  Dinner     3        0.17
3         23.68 3.31    Male     No   Sun  Dinner     2        0.14
4         24.59 3.61  Female     No   Sun  Dinner     4        0.15
..          ...  ...     ...    ...   ...     ...   ...         ...
239       29.03 5.92    Male     No   Sat  Dinner     3        0.20
240       27.18 2.00  Female    Yes   Sat  Dinner     2        0.07
241       22.67 2.00    Male    Yes   Sat  Dinner     2        0.09
242       17.82 1.75    Male     No   Sat  Dinner     2        0.10
243       18.78 3.00  Female     No  Thur  Dinner     2        0.16

[244 rows x 8 columns]

In [52]:
df.groupby('sex').mean()

total_bill  tip  size  porcet_tip
sex                                      
Male         20.74 3.09  2.63        0.16
Female       18.06 2.83  2.46        0.17

La porcentaje de propina en las mujeres respecto al total de su cuenta es mayor que el de los hombres. Con esto se puede concluir que las mujeres dejan mayores propinas. 

Podemos utilizar la función median() que tiene en cuenta una distribución normal de los datos y podemos estimar un porcentaje de propina mas aproximado a lo real. 

In [53]:
df.groupby('sex').median()

total_bill  tip  size  porcet_tip
sex                                      
Male         18.35 3.00     2        0.15
Female       16.40 2.75     2        0.16

Tambien podemos ver estimadores estadisticos por agrupaciones. En este caso agrupamos por **sex** y luego describimos estadisticamente la bajo la columna **porcet_tip**.

In [54]:
df.groupby('sex')[['porcet_tip']].describe()

porcet_tip                                   
            count mean  std  min  25%  50%  75%  max
sex                                                 
Male       157.00 0.16 0.06 0.04 0.12 0.15 0.19 0.71
Female      87.00 0.17 0.05 0.06 0.14 0.16 0.19 0.42

O lo podemos hacer con dos culumnas.

In [55]:
df.groupby('sex')[['total_bill','porcet_tip']].describe()

total_bill                              ... porcet_tip                         
            count  mean  std  min   25%   50%  ...        std  min  25%  50%  75%  max
sex                                            ...                                    
Male       157.00 20.74 9.25 7.25 14.00 18.35  ...       0.06 0.04 0.12 0.15 0.19 0.71
Female      87.00 18.06 8.01 3.07 12.75 16.40  ...       0.05 0.06 0.14 0.16 0.19 0.42

[2 rows x 16 columns]

Creemoa una funcion para pasar de euros a dolares las cuenta, asumiendo que todo se encuentra en euros.

In [56]:
def mean_eur2usd(x):
  return np.mean(x) * 1.12

In [57]:
mean_eur2usd(100)

112.00000000000001

Podemos aplicar nuestra función a un solo grupo creado por nosotros o a varios grupos. 

In [58]:
df.groupby('sex')[['total_bill','porcet_tip']].apply(mean_eur2usd)

total_bill  porcet_tip
sex                           
Male         23.23        0.18
Female       20.22        0.19

In [59]:
df.groupby(['sex','time'])[['total_bill','porcet_tip']].apply(mean_eur2usd)

total_bill  porcet_tip
sex    time                          
Male   Lunch        20.21        0.19
       Dinner       24.04        0.17
Female Lunch        18.30        0.18
       Dinner       21.52        0.19

Podemos realizar lo mismo que la funcion creada pero con lambdas.

In [60]:
df.groupby(['sex','time'])[['total_bill','porcet_tip']].apply(lambda x:np.mean(x) * 1.12)

total_bill  porcet_tip
sex    time                          
Male   Lunch        20.21        0.19
       Dinner       24.04        0.17
Female Lunch        18.30        0.18
       Dinner       21.52        0.19

Podemos calcular la desviación estadar sobre nuestro grupo.

In [61]:
df.groupby(['sex','time'])[['total_bill','porcet_tip']].apply(np.std)

total_bill  porcet_tip
sex    time                          
Male   Lunch         7.83        0.05
       Dinner        9.42        0.07
Female Lunch         7.39        0.03
       Dinner        8.12        0.06

Si quisieramos utilizar mas de una funcion podemos utilizar aggregate

In [62]:
df.groupby(['sex','time'])[['total_bill','porcet_tip']].aggregate([np.mean, np.max])

total_bill       porcet_tip     
                    mean  amax       mean amax
sex    time                                   
Male   Lunch       18.05 41.19       0.17 0.27
       Dinner      21.46 50.81       0.16 0.71
Female Lunch       16.34 43.11       0.16 0.26
       Dinner      19.21 44.30       0.17 0.42

Utilizando aggregate podemos tambien agregar diferentes funciones, tambien podemos crear un diccionario para agregar difenretes funciones por categoria, como por ejemplo:

In [63]:
dict_agg = {
    'tip':[min, max],
    'total_bill':[np.mean, mean_eur2usd]
}

dict_agg

{'tip': [<function min>, <function max>],
 'total_bill': [<function numpy.mean>, <function __main__.mean_eur2usd>]}

In [64]:
df.groupby(['sex','time'])[['total_bill','tip']].agg(dict_agg)

tip       total_bill             
               min   max       mean mean_eur2usd
sex    time                                     
Male   Lunch  1.44  6.70      18.05        20.21
       Dinner 1.00 10.00      21.46        24.04
Female Lunch  1.25  5.17      16.34        18.30
       Dinner 1.00  6.50      19.21        21.52

Podemos crear una funcion filtro y aplicarla al group by creado por nosotros

In [65]:
def f_filter(x):
  return mean_eur2usd(x['total_bill'].mean()) > 20

In [66]:
 df_filtered = df.groupby(['sex', 'time']).filter(f_filter)
 df_filtered

total_bill  tip     sex smoker   day    time  size  porcet_tip
0         16.99 1.01  Female     No   Sun  Dinner     2        0.06
1         10.34 1.66    Male     No   Sun  Dinner     3        0.16
2         21.01 3.50    Male     No   Sun  Dinner     3        0.17
3         23.68 3.31    Male     No   Sun  Dinner     2        0.14
4         24.59 3.61  Female     No   Sun  Dinner     4        0.15
..          ...  ...     ...    ...   ...     ...   ...         ...
239       29.03 5.92    Male     No   Sat  Dinner     3        0.20
240       27.18 2.00  Female    Yes   Sat  Dinner     2        0.07
241       22.67 2.00    Male    Yes   Sat  Dinner     2        0.09
242       17.82 1.75    Male     No   Sat  Dinner     2        0.10
243       18.78 3.00  Female     No  Thur  Dinner     2        0.16

[209 rows x 8 columns]

En el dataset original teniamos **244 rows**, pero usando el filtro creado ahora tenemos **209 rows**. Esto quiere decir que lo que estamos haciendo es eliminar todas las filas donde el costo promedio de la variable **total_bill** sea mayor a 20.

In [67]:
df_filtered.groupby(['sex', 'time']).count()

total_bill    tip  smoker    day   size  porcet_tip
sex    time                                                       
Male   Lunch        33.00  33.00   33.00  33.00  33.00       33.00
       Dinner      124.00 124.00  124.00 124.00 124.00      124.00
Female Lunch          nan    nan     nan    nan    nan         nan
       Dinner       52.00  52.00   52.00  52.00  52.00       52.00

# Group By: extraer valor con variables categóricas

En la seccion anterior se aprendio como **extraer valor** de las **variables numericas** con las funcion groupby y aggregate, usando categorias. 

En esta seccion se vera como **extraer valor** de los datos  de las **variables categoricas**.

Veamos nuevamente nuestro dataframe:

In [68]:
df

total_bill  tip     sex smoker   day    time  size  porcet_tip
0         16.99 1.01  Female     No   Sun  Dinner     2        0.06
1         10.34 1.66    Male     No   Sun  Dinner     3        0.16
2         21.01 3.50    Male     No   Sun  Dinner     3        0.17
3         23.68 3.31    Male     No   Sun  Dinner     2        0.14
4         24.59 3.61  Female     No   Sun  Dinner     4        0.15
..          ...  ...     ...    ...   ...     ...   ...         ...
239       29.03 5.92    Male     No   Sat  Dinner     3        0.20
240       27.18 2.00  Female    Yes   Sat  Dinner     2        0.07
241       22.67 2.00    Male    Yes   Sat  Dinner     2        0.09
242       17.82 1.75    Male     No   Sat  Dinner     2        0.10
243       18.78 3.00  Female     No  Thur  Dinner     2        0.16

[244 rows x 8 columns]

Creemos una nueva columna llamada **ones** a la cual le pondremos numeros uno unicamente.

In [69]:
df['ones'] = 1

In [70]:
df

total_bill  tip     sex smoker   day    time  size  porcet_tip  ones
0         16.99 1.01  Female     No   Sun  Dinner     2        0.06     1
1         10.34 1.66    Male     No   Sun  Dinner     3        0.16     1
2         21.01 3.50    Male     No   Sun  Dinner     3        0.17     1
3         23.68 3.31    Male     No   Sun  Dinner     2        0.14     1
4         24.59 3.61  Female     No   Sun  Dinner     4        0.15     1
..          ...  ...     ...    ...   ...     ...   ...         ...   ...
239       29.03 5.92    Male     No   Sat  Dinner     3        0.20     1
240       27.18 2.00  Female    Yes   Sat  Dinner     2        0.07     1
241       22.67 2.00    Male    Yes   Sat  Dinner     2        0.09     1
242       17.82 1.75    Male     No   Sat  Dinner     2        0.10     1
243       18.78 3.00  Female     No  Thur  Dinner     2        0.16     1

[244 rows x 9 columns]

Con lo anterior, creemos un nuevo dataframe donde tomemos las **categorias** **sexo** y **fumadores**, luego utilizando la columna antes creada hagamos un **conteo** de los mismos separandolos por categorias. 

In [71]:
df_g = df.groupby(['sex', 'smoker'])['ones'].sum()
df_g

sex     smoker
Male    Yes       60
        No        97
Female  Yes       33
        No        54
Name: ones, dtype: int64

Con esto hicimos un conteo y un analisis con las variables categoricas, las cuales no son numericas, valga la redundancia.

Ahora si de lo anterior quisieramos convertirlo en un % lo que hacemos es lo siguiente:

In [72]:
df_g.groupby(level=0).apply(
    lambda x:
      x / x.sum() * 100
)

sex     smoker
Male    Yes      38.22
        No       61.78
Female  Yes      37.93
        No       62.07
Name: ones, dtype: float64

Otra forma de crear valor de nuestros datos es transformar alguna variable numerica (continua) en categorica, ¿como hacemos esto?

In [73]:
pd.cut(df['total_bill'], bins=3)

0       (3.022, 18.983]
1       (3.022, 18.983]
2      (18.983, 34.897]
3      (18.983, 34.897]
4      (18.983, 34.897]
             ...       
239    (18.983, 34.897]
240    (18.983, 34.897]
241    (18.983, 34.897]
242     (3.022, 18.983]
243     (3.022, 18.983]
Name: total_bill, Length: 244, dtype: category
Categories (3, interval[float64]): [(3.022, 18.983] < (18.983, 34.897] < (34.897, 50.81]]

Con esto hemos creado categorias con variables numericas. Ahora veamos cuales fueron las categorias creadas:

In [74]:
pd.cut(df['total_bill'], bins=3).value_counts()

(3.022, 18.983]     140
(18.983, 34.897]     88
(34.897, 50.81]      16
Name: total_bill, dtype: int64

Tambien podemos definir el ancho de nuestras categorias:

In [75]:
pd.cut(df['total_bill'], bins=[3,18,35,60]).value_counts()

(3, 18]     127
(18, 35]    101
(35, 60]     16
Name: total_bill, dtype: int64

Ahora que creamos categorias de las variables numericas almacenemos lo hecho en nuestro datarame:

In [76]:
df['bin_total'] = pd.cut(df['total_bill'], bins=[3,18,35,60])
df

total_bill  tip     sex smoker  ... size porcet_tip  ones  bin_total
0         16.99 1.01  Female     No  ...    2       0.06     1    (3, 18]
1         10.34 1.66    Male     No  ...    3       0.16     1    (3, 18]
2         21.01 3.50    Male     No  ...    3       0.17     1   (18, 35]
3         23.68 3.31    Male     No  ...    2       0.14     1   (18, 35]
4         24.59 3.61  Female     No  ...    4       0.15     1   (18, 35]
..          ...  ...     ...    ...  ...  ...        ...   ...        ...
239       29.03 5.92    Male     No  ...    3       0.20     1   (18, 35]
240       27.18 2.00  Female    Yes  ...    2       0.07     1   (18, 35]
241       22.67 2.00    Male    Yes  ...    2       0.09     1   (18, 35]
242       17.82 1.75    Male     No  ...    2       0.10     1    (3, 18]
243       18.78 3.00  Female     No  ...    2       0.16     1   (18, 35]

[244 rows x 10 columns]

Ahora sobre esta nueva categoria podemos hacer group by. En este caso vamos a contar en funcion de la categoria antes hecha **bin_total** en que horario, **time**, se gasta mas o menos. 

In [77]:
df.groupby(['time', 'bin_total'])[['ones']].count()

ones
time   bin_total      
Lunch  (3, 18]      45
       (18, 35]     21
       (35, 60]      2
Dinner (3, 18]      82
       (18, 35]     80
       (35, 60]     14

Veamos lo anterior en %

In [78]:
df.groupby(['time', 'bin_total'])[['ones']].count().groupby(level=0).apply(
    lambda x:
      x / x.sum() * 100
)

ones
time   bin_total      
Lunch  (3, 18]   66.18
       (18, 35]  30.88
       (35, 60]   2.94
Dinner (3, 18]   46.59
       (18, 35]  45.45
       (35, 60]   7.95

Vemos que en el almuerzo la gente suele pagar facturas mas baratas ya que tenemos un 66.18% de estas en el almuerzo y sobre las facturas de mayor valor, entre 35 y 60 USD solo tenemos un 2.94%.

En cambio en la noche, en la cena, esto % cambian. Las personas puede pagar mas por una comina ya que los % de la categoria entre 18-35 USD y 35-60 USD aumentan, haciendo que la primer categoria disminuya. 

Con esto podemos concluir que la gente suele gastar mas en una cena que en una almuerzo. 

# Tablas dinámicas con Pivot Table

En esta sección veremos como crear tablas dinámicas. Para eso pandas cuenta con una función llamada **pivot_table**, esta nos permite extraer, cuando tenemos variables categoricas, gran informacion y valor de nuestro dataframe.

In [79]:
df

total_bill  tip     sex smoker  ... size porcet_tip  ones  bin_total
0         16.99 1.01  Female     No  ...    2       0.06     1    (3, 18]
1         10.34 1.66    Male     No  ...    3       0.16     1    (3, 18]
2         21.01 3.50    Male     No  ...    3       0.17     1   (18, 35]
3         23.68 3.31    Male     No  ...    2       0.14     1   (18, 35]
4         24.59 3.61  Female     No  ...    4       0.15     1   (18, 35]
..          ...  ...     ...    ...  ...  ...        ...   ...        ...
239       29.03 5.92    Male     No  ...    3       0.20     1   (18, 35]
240       27.18 2.00  Female    Yes  ...    2       0.07     1   (18, 35]
241       22.67 2.00    Male    Yes  ...    2       0.09     1   (18, 35]
242       17.82 1.75    Male     No  ...    2       0.10     1    (3, 18]
243       18.78 3.00  Female     No  ...    2       0.16     1   (18, 35]

[244 rows x 10 columns]

In [80]:
df.groupby(['sex', 'time'])[['total_bill']].mean()

total_bill
sex    time              
Male   Lunch        18.05
       Dinner       21.46
Female Lunch        16.34
       Dinner       19.21

Lo que haremos es 

In [81]:
df_gp = df.groupby(['sex', 'time'])[['total_bill']].mean().reset_index()
df_gp

sex    time  total_bill
0    Male   Lunch       18.05
1    Male  Dinner       21.46
2  Female   Lunch       16.34
3  Female  Dinner       19.21

In [85]:
df_gp.pivot_table(index='sex', columns='time', values='total_bill')

time    Lunch  Dinner
sex                  
Male    18.05   21.46
Female  16.34   19.21

Que esta pasando con **pivot_table**?

La tabla se reestructuran como podemos ver en la imagen:

![](https://pandas-docs.github.io/pandas-docs-travis/_images/reshaping_pivot.png)

Internamente **pivot_table** esta realizando el promedio de cada uno de los valores y reestructura a nuestro dataframe en uno nuevo con lo hemos especificado.

In [86]:
df.pivot_table(index='sex', columns='time', values='total_bill')

time    Lunch  Dinner
sex                  
Male    18.05   21.46
Female  16.34   19.21

Si quisieramos cambiar el tipo de cálculo que hace este lo podemos hacer con el atributo **aggfunc** y especificarle la función.

In [87]:
df.pivot_table(index='sex', columns='time', values='total_bill', aggfunc=np.median)

time    Lunch  Dinner
sex                  
Male    16.58   19.63
Female  13.42   17.19

Tambien podemos pasarla una lista con las funciones que queremos utilizar para que haga los calculos internos la funcion **pivot_table**. En este caso la media y la desviación estandar.

In [89]:
df_pivot = df.pivot_table(index='sex', columns='time', values='total_bill', aggfunc=[np.median, np.std])
df_pivot

median          std       
time    Lunch Dinner Lunch Dinner
sex                              
Male    16.58  19.63  7.95   9.46
Female  13.42  17.19  7.50   8.20

Podemos deshacer las categorias y crear un nuevo dataframe de la siguiente manera.

In [90]:
df_pivot.unstack()

        time    sex   
median  Lunch   Male     16.58
                Female   13.42
        Dinner  Male     19.63
                Female   17.19
std     Lunch   Male      7.95
                Female    7.50
        Dinner  Male      9.46
                Female    8.20
dtype: float64

In [91]:
df_pivot.unstack().reset_index()

level_0    time     sex     0
0  median   Lunch    Male 16.58
1  median   Lunch  Female 13.42
2  median  Dinner    Male 19.63
3  median  Dinner  Female 17.19
4     std   Lunch    Male  7.95
5     std   Lunch  Female  7.50
6     std  Dinner    Male  9.46
7     std  Dinner  Female  8.20